src/esp/gfx/PTexMeshDrawable.cpp

```diff
@@ -18,18 +20,45 @@ PTexMeshDrawable::PTexMeshDrawable(
     : Drawable{node, shader, ptexMeshData.getRenderingBuffer(submeshID)->mesh,
                group},
       tex_(ptexMeshData.getRenderingBuffer(submeshID)->tex),
-      adjTex_(ptexMeshData.getRenderingBuffer(submeshID)->adjTex),
+      adjFaces_(ptexMeshData.getRenderingBuffer(submeshID)->adjFaces),
       tileSize_(ptexMeshData.tileSize()),
-      exposure_(ptexMeshData.exposure()) {}
+      exposure_(ptexMeshData.exposure()),
+      gamma_(ptexMeshData.gamma()),
+      saturation_(ptexMeshData.saturation()) {
+  PTexMeshShader& ptexMeshShader = static_cast<PTexMeshShader&>(shader_);
+  // clipPlane is const for every ptex mesh in this version.
+  // Set it in the constructor, not in the draw()
+  ptexMeshShader.setClipPlane(clipPlane_);
+}
 
 void PTexMeshDrawable::draw(const Magnum::Matrix4& transformationMatrix,
                             Magnum::SceneGraph::Camera3D& camera) {
-  adjTex_.bind(1);
   PTexMeshShader& ptexMeshShader = static_cast<PTexMeshShader&>(shader_);
-  ptexMeshShader.bindTexture(tex_, 0)
-      .setPTexUniforms(tex_, tileSize_, exposure_)
+  ptexMeshShader.setExposure(exposure_)
+      .setGamma(gamma_)
+      .setSaturation(saturation_)
+      .setAtlasTextureSize(tex_, tileSize_)
+      .bindAtlasTexture(tex_)
+      .bindAdjFacesBufferTexture(adjFaces_)
       .setMVPMatrix(camera.projectionMatrix() * transformationMatrix);
   mesh_.draw(ptexMeshShader);
+
+  //XXX
+  // ASSERT(Magnum::GL::Renderer::Error::NoError == Magnum::GL::Renderer::error());
+  switch (Magnum::GL::Renderer::error()) {
```

Comment:
Does the InvalidOperation happen on Linux too? 

- If not, then that is a potential source of the problem :) 
- If yes, can you run on Linux with `--magnum-gpu-validation on` passed to the application? (Or setting the `MAGNUM_GPU_VALIDATION` environment variable to `ON`, in case you can't pass arguments.) That should tell you the origin of the warning in the console output. Ideally code that works shouldn't be emitting any errors.